code/xcvm/evm/src/Gateway.sol

```diff
@@ -0,0 +1,145 @@
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "./Interpreter.sol";
+import "./interfaces/IGateway.sol";
+
+contract Gateway is Ownable, IGateway {
+    mapping(uint256 => mapping(bytes => address)) public userInterpreter;
+
+    mapping(address => Bridge) public bridgesInfo;
+    // TODO ? do we have only one bridgge per network and security
+    mapping(uint256 => mapping(BridgeSecurity => address)) public bridges;
+    mapping(uint256 => address) public assets;
+
+    event InstanceCreated(uint256 networkId, bytes user, address instance);
+
+    event Spawn(
+        bytes account,
+        uint256 networkId,
+        BridgeSecurity security,
+        uint256 salt,
+        bytes spawnedProgram,
+        address[] assetAddresses,
+        uint256[] amounts
+    );
+
+    modifier onlyBridge() {
+        require(bridgesInfo[msg.sender].security != BridgeSecurity(0));
+        _;
+    }
+
+    constructor() {
+        // enable trustless bridge;
+    }
+
+    function getAsset(uint256 assetId) external returns (address) {
+        return assets[assetId];
+    }
+
+    function getBridge(uint256 networkId, BridgeSecurity security) external returns (address) {
+        return bridges[networkId][security];
+    }
+
+    function registerAsset(address assetAddress, uint128 assetId) external onlyOwner {
+        require(assetAddress != address(0), "Gateway: invalid address");
+        assets[assetId] = assetAddress;
+    }
+
+    function unregisterAsset(uint128 assetId) external onlyOwner {
+        delete assets[assetId];
+    }
+
+    function registerBridge(
+        address bridgeAddress,
+        BridgeSecurity security,
+        uint256 networkId
+    ) external onlyOwner {
+        require(bridges[networkId][security] == address(0), "Gateway: this type of bridge already registered");
+        require(bridgeAddress != address(0), "Gateway: invalid address");
+        require(bridgesInfo[bridgeAddress].security == BridgeSecurity(0), "Gateway: bridge already enabled");
+        require(security != BridgeSecurity(0), "Gateway: should not disable bridge while registering bridge");
+        bridgesInfo[bridgeAddress].security = security;
+        bridgesInfo[bridgeAddress].networkId = networkId;
+        bridges[networkId][security] = bridgeAddress;
+    }
+
+    function unregisterBridge(address bridgeAddress) external onlyOwner {
+        require(bridgesInfo[bridgeAddress].security != BridgeSecurity(0), "Gateway: bridge already disabled");
+        delete bridges[bridgesInfo[bridgeAddress].networkId][bridgesInfo[bridgeAddress].security];
+        bridgesInfo[bridgeAddress].security = BridgeSecurity(0);
+        bridgesInfo[bridgeAddress].networkId = 0;
+    }
+
+    //// TODO ? is the bridge who's gonna to provide internetwork assets transfer?
+    function _provisionAssets(
+        address payable interpreterAddress,
+        address[] memory erc20AssetList,
```

Comment:
If I am not wrong, if the data is read-only, using `callback` instead of `memory` uses less gas. (This applies to other places that uses `memory` in read-only case)

Reply:
gotcha thanks